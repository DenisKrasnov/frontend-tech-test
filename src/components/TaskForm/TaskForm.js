import React, { Component } from "react";
import { shape, string, bool, func } from "prop-types";

import { errorMessages, maxLengths } from "../../utilities/utilities";
import Button from "../Button";
import Counter from "../Counter";
import Notification from "../Notification";
import styles from "./TaskForm.scss";

class TaskForm extends Component {
  static propTypes = {
    task: shape({
      title: string,
      description: string,
    }),
    headerName: string,
    onSubmitName: string.isRequired,
    onSubmit: func.isRequired,
    loading: bool,
    validate: func,
  };

  static defaultProps = {
    task: { title: "", description: "" },
    headerName: "",
    loading: false,
    validate: () => {},
  };

  state = {
    title: "",
    description: "",
    validationError: "",
  };

  componentDidMount() {
    const { task } = this.props;
    if (task) {
      this.setState({ title: task.title, description: task.description });
    }
  }

  setErrorMessage = message => this.setState({ validationError: message });

  validate = ({ title, description }) => {
    const { validate } = this.props;

    if (validate) {
      const errorMessage = validate();
      if (errorMessage) {
        this.setErrorMessage(errorMessage);
        return null;
      }
    }

    const cleanTitle = title.trim();
    const cleanDescription = description.trim();

    if (cleanTitle === "" && cleanDescription === "") {
      this.setErrorMessage(errorMessages.emptyFields);
      return null;
    }

    return {
      title: cleanTitle,
      description: cleanDescription,
    };
  };

  onSubmit = e => {
    e.preventDefault();

    const { onSubmit } = this.props;
    const { title, description } = this.state;
    const validValues = this.validate({ title, description });

    if (validValues) {
      onSubmit(validValues);
      this.clearState();
    }
  };

  onChange = e => {
    const { name, value } = e.target;
    this.setState({
      [name]: value,
    });
  };

  clearState = () =>
    this.setState({
      title: "",
      description: "",
      validationError: "",
    });

  hideNotification = () => {
    this.setState({
      validationError: "",
    });
  };

  render() {
    const { loading, headerName, onSubmitName } = this.props;
    const { title, description, validationError } = this.state;

    return (
      <div className={styles.TaskForm}>
        {headerName ? (
          <div className={styles.header}>
            <h2 className={styles.name}>{headerName}</h2>
          </div>
        ) : null}

        {validationError ? (
          <Notification onDismiss={this.hideNotification} type="error">
            {validationError}
          </Notification>
        ) : null}
        <form acceptCharset="utf-8" onSubmit={this.onSubmit}>
          <label className={styles.field}>
            <Counter maxLength={maxLengths.title} valueLength={title.length} />
            Title:
            <input
              className={styles.input}
              maxLength={maxLengths.title}
              name="title"
              onChange={this.onChange}
              type="text"
              value={title}
            />
          </label>
          <label className={styles.field}>
            <Counter
              maxLength={maxLengths.description}
              valueLength={description.length}
            />
            Description:
            <textarea
              className={styles.textarea}
              maxLength={maxLengths.description}
              name="description"
              onChange={this.onChange}
              rows="3"
              value={description}
            />
          </label>
          <Button loading={loading}>{onSubmitName}</Button>
        </form>
      </div>
    );
  }
}

export default TaskForm;
