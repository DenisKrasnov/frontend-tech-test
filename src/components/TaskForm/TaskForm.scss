.TaskForm {
  display: flex;
  flex-direction: column;
}

.header {
  padding: 0 0 25px;
  text-align: left;
}

.name {
  font-weight: 600;
  display: inline;
}

.field {
  display: block;
  margin-bottom: 10px;
  font-weight: 600;
}

.input {
  display: block;
  width: 100%;
  min-height: 42px;
  padding: 10px 16px;
  margin-top: 10px;
  border: 1px solid $border-color;
  outline: 0;
  background: #fff;
  color: #12438a;
  line-height: 20px;

  &:focus {
    border-color: $border-dark-color;
  }
}

.textarea {
  composes: input;
  resize: vertical;
}
