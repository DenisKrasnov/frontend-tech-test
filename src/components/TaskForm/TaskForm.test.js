import React from "react";
import { shallow } from "enzyme";

import TaskForm from "./TaskForm";

const task = {
  id: 2,
  title: "__TITLE__",
  description: "__DESCRIPTION__",
};

describe("TaskForm", () => {
  it("renders", () => {
    const component = shallow(
      <TaskForm
        onSubmitName="Create task"
        headerName="Create Task"
        onSubmit={() => {}}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it("renders with initial values when the 'task' is passed in", () => {
    const component = shallow(
      <TaskForm
        onSubmitName="Create task"
        headerName="Create Task"
        onSubmit={() => {}}
        task={task}
      />
    );

    expect(component.state().title).toEqual(task.title);
    expect(component.state().description).toEqual(task.description);

    expect(component).toMatchSnapshot();
  });

  it("renders the error notification when the values are empty", () => {
    const task = {
      id: 2,
      title: "",
      description: "",
    };
    const component = shallow(
      <TaskForm
        onSubmitName="Create task"
        headerName="Create Task"
        onSubmit={() => {}}
        task={task}
      />
    );

    component.instance().onSubmit({ preventDefault: () => {} });

    expect(component.state().validationError).not.toBeFalsy();
    expect(component).toMatchSnapshot();
  });

  it("triggers onSubmit prop function and clears the state", () => {
    const onSubmitSpy = jest.fn();
    const component = shallow(
      <TaskForm
        onSubmitName="Create task"
        headerName="Create Task"
        onSubmit={onSubmitSpy}
        task={task}
      />
    );

    component.instance().onSubmit({ preventDefault: () => {} });
    expect(onSubmitSpy).toHaveBeenCalledWith({
      title: task.title,
      description: task.description,
    });

    expect(component.state().title).toBe("");
    expect(component.state().description).toBe("");
    expect(component.state().validationError).toBe("");
  });
});
