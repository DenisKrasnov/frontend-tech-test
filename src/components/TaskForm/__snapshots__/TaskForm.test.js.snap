// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskForm renders 1`] = `
<div
  className="TaskForm"
>
  <div
    className="header"
  >
    <h2
      className="name"
    >
      Create Task
    </h2>
  </div>
  <form
    acceptCharset="utf-8"
    onSubmit={[Function]}
  >
    <label
      className="field"
    >
      <Counter
        maxLength={40}
        valueLength={0}
      />
      Title:
      <input
        className="input"
        maxLength={40}
        name="title"
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
    <label
      className="field"
    >
      <Counter
        maxLength={300}
        valueLength={0}
      />
      Description:
      <textarea
        className="textarea"
        maxLength={300}
        name="description"
        onChange={[Function]}
        rows="3"
        value=""
      />
    </label>
    <Button
      loading={false}
      onClick={[Function]}
      type="submit"
    >
      Create task
    </Button>
  </form>
</div>
`;

exports[`TaskForm renders the error notification when the values are empty 1`] = `
<div
  className="TaskForm"
>
  <div
    className="header"
  >
    <h2
      className="name"
    >
      Create Task
    </h2>
  </div>
  <Notification
    onDismiss={[Function]}
    type="error"
  >
    Please complete at least one of the fields
  </Notification>
  <form
    acceptCharset="utf-8"
    onSubmit={[Function]}
  >
    <label
      className="field"
    >
      <Counter
        maxLength={40}
        valueLength={0}
      />
      Title:
      <input
        className="input"
        maxLength={40}
        name="title"
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
    <label
      className="field"
    >
      <Counter
        maxLength={300}
        valueLength={0}
      />
      Description:
      <textarea
        className="textarea"
        maxLength={300}
        name="description"
        onChange={[Function]}
        rows="3"
        value=""
      />
    </label>
    <Button
      loading={false}
      onClick={[Function]}
      type="submit"
    >
      Create task
    </Button>
  </form>
</div>
`;

exports[`TaskForm renders with initial values when the 'task' is passed in 1`] = `
<div
  className="TaskForm"
>
  <div
    className="header"
  >
    <h2
      className="name"
    >
      Create Task
    </h2>
  </div>
  <form
    acceptCharset="utf-8"
    onSubmit={[Function]}
  >
    <label
      className="field"
    >
      <Counter
        maxLength={40}
        valueLength={9}
      />
      Title:
      <input
        className="input"
        maxLength={40}
        name="title"
        onChange={[Function]}
        type="text"
        value="__TITLE__"
      />
    </label>
    <label
      className="field"
    >
      <Counter
        maxLength={300}
        valueLength={15}
      />
      Description:
      <textarea
        className="textarea"
        maxLength={300}
        name="description"
        onChange={[Function]}
        rows="3"
        value="__DESCRIPTION__"
      />
    </label>
    <Button
      loading={false}
      onClick={[Function]}
      type="submit"
    >
      Create task
    </Button>
  </form>
</div>
`;
